src/nvim/lib/rbuffer.h

```diff
@@ -0,0 +1,285 @@
+// Ring buffer implementation. This is basically an array that wraps read/write
+// pointers around the memory region. It should be more efficient than the old
+// RBuffer which required memmove() calls to relocate read/write positions.
+//
+// The main purpose of RBuffer is simplify memory management when reading from
+// uv_stream_t instances:
+//
+// - The event loop writes data to a RBuffer, advancing the write pointer
+// - The main loop reads data, advancing the read pointer
+// - If the buffer becomes full(size == capacity) the rstream is temporarily
+//   stopped(automatic backpressure handling)
+//
+// Reference: http://en.wikipedia.org/wiki/Circular_buffer
+#ifndef NVIM_LIB_RBUFFER_H
+#define NVIM_LIB_RBUFFER_H
+
+#include <assert.h>
+#include <stddef.h>
+#include <string.h>
+
+#include "nvim/memory.h"
+#include "nvim/vim.h"
+
+
+// Macros that simplify working with the read/write pointers directly by hiding
+// ring buffer wrap logic. Some examples:
+//
+// - Pass the write pointer to a function(`write_data`) that incrementally
+//   produces data, returning the number of bytes actually written:
+//
+//       RBUFFER_EACH_WRITE_PTR(rbuf, ptr, cnt)
+//         rbuffer_produced(rbuf, write_data(state, ptr, cnt));
+//
+// - Pass the read pointer to a function(`read_data`) that incrementally
+//   consumes data, return the number of bytes actually read:
+//
+//       RBUFFER_EACH_READ_PTR(rbuf, ptr, cnt)
+//         rbuffer_consumed(rbuf, read_data(state, ptr, cnt));
+//
+// In other words, these functions iterate through each read/write pointers
+// passing their values to the `ptr` argument, while the `cnt` argument contains
+// the real size of the buffer pointed by `ptr`(the maximum number of bytes that
+// can be read or written). `rbuffer_produced`/`rbuffer_consumed` are the
+// functions that actually mutate the internal read/write pointers to reflect
+// read/written data.
+#define RBUFFER_EACH_READ_PTR(buf, rptr, rcnt)                                 \
+  for (size_t rcnt = 0, _r = buf->size; _r != SIZE_MAX; _r = SIZE_MAX)         \
+    for (char *rptr = rbuffer_read_ptr(buf, &rcnt);                            \
+         _r;                                                                   \
+         _r -= rcnt, rptr = _r ? buf->start_ptr : NULL, rcnt = _r ? _r : 0)
```

Comment:
Would it be problematic to use the macro with an additional (code) argument? Seems to be cleaner, easier to check and does not have the problem with reads that consume less than `rcnt` bytes in one iteration (from the example above `RBUFFER_EACH_READ_PTR(rbuf, ptr, cnt) rbuffer_consumed(rbuf, read_data(state, ptr, cnt));`).
```c
#define RBUFFER_EACH_READ_PTR(buf, rptr, rcnt, code)                          \
  do {                                                                         \
    size_t rcnt;                                                               \
    while (buf->size) {                                                        \
      char *rptr = rbuffer_read_ptr(buf, &rcnt);                               \
      code                                                                     \
    }                                                                          \
  } while (0)
```
Usage is similar to the old macro:
```c
static inline size_t rbuffer_read(RBuffer *buf, char *dst, size_t dst_size)
  FUNC_ATTR_NONNULL_ALL
{
  size_t size = dst_size;
  RBUFFER_EACH_READ_PTR(buf, rptr, rcnt, {
    size_t copy_count = MIN(dst_size, rcnt);
    memcpy(dst, rptr, copy_count);
    rbuffer_consumed(buf, copy_count);

    if (!(dst_size -= copy_count)) {
      return size;
    }

    dst += copy_count;
  });

  return size - dst_size;
}
```